processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go

```diff
@@ -0,0 +1,56 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ec2
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+)
+
+const (
+	TypeStr          = "ec2"
+	cloudProviderAWS = "aws"
+)
+
+type Detector struct {
+	provider ec2MetadataProvider
+}
+
+func NewDetector() *Detector {
+	return &Detector{provider: ec2MetadataImpl{}}
+}
+
+func (d *Detector) Detect(ctx context.Context) (pdata.Resource, error) {
+	res := pdata.NewResource()
+	res.InitEmpty()
+	attr := res.Attributes()
+	attr.InsertString(conventions.AttributeCloudProvider, cloudProviderAWS)
```

Comment:
Should this be set after getting the metadata? For instance if `d.provide.get` failed because the current resource is not on AWS, we should probably not set this field.

Reply:
You're right, for some reason I thought the GCE detector did that but it's not. Fixed.